zio-kafka/src/main/scala/zio/kafka/consumer/internal/Runloop.scala

```diff
@@ -57,113 +62,152 @@ private[consumer] final class Runloop private (
+          _ <- ZIO.foreachDiscard(lostStreams)(_.lost())
+          streamsToEnd = if (restartStreamsOnRebalancing && !rebalanceEvent.wasInvoked) remainingStreams
+                         else Chunk.empty
+          pendingCommits <- endRevokedStreamsAndAwaitEnd(rebalanceConsumer, streamsToEnd)
+          _              <- rebalanceListenerEvent.update(_.onRevokedOrLost(pendingCommits))
+          _              <- ZIO.logTrace(s"onLost done")
+        } yield ()
     )
 
-    if (restartStreamsOnRebalancing) {
-      emitDiagnostics ++ restartStreamsRebalancingListener ++ userRebalanceListener
-    } else {
-      emitDiagnostics ++ userRebalanceListener
-    }
+    emitDiagnostics ++ endRevokedStreamsRebalancingListener ++ userRebalanceListener
   }
 
-  private val commit: Map[TopicPartition, Long] => Task[Unit] =
-    offsets =>
-      for {
-        p <- Promise.make[Throwable, Unit]
-        _ <- commandQueue.offer(Commit(offsets, p)).unit
-        _ <- diagnostics.emitIfEnabled(DiagnosticEvent.Commit.Started(offsets))
-        _ <- p.await
-      } yield ()
+  private def endRevokedStreamsAndAwaitEnd(
+    rebalanceConsumer: RebalanceConsumer,
+    streamsToEnd: Chunk[PartitionStreamControl]
+  ): Task[Chunk[Commit]] =
+    for {
+      _ <- ZIO.foreachDiscard(streamsToEnd)(_.end())
+      pendingCommits <-
+        if (streamsToEnd.nonEmpty && endRevokedStreamsBeforeRebalance) {
+          // When the queue is empty we still need to call commit (with 0 offsets) so that earlier commits can complete.
+          // We can not use ZStream.fromQueue because that will emit nothing when the queue is empty.
+          // TODO: add timeout; we can't postpone the rebalance for ever
+          ZStream
+            .fromZIO(commitQueue.takeAll)
+            .repeat(Schedule.forever)
+            .tap(doCommitsFromRebalanceListener(rebalanceConsumer))
+            .takeUntilZIO(_ => ZIO.forall(streamsToEnd)(_.isCompleted))
+            .runCollect
+            .map(_.flatten)
+        } else {
+          ZIO.succeed(Chunk.empty)
+        }
+    } yield pendingCommits
+
+  /**
+   * Handle commits while waiting for revoked streams to end.
+   *
+   * We need to ensure the revoked streams can end. This is only possible if the commits these streams started complete.
+   * The commits complete when the callback is invoked. The callback is invoked when the underlying consumer polls the
+   * broker. This can be achieved by invoking `commitAsync`. Even when we pass no offsets, the broker will be polled and
+   * callbacks will be called.
+   */
+  private def doCommitsFromRebalanceListener(
+    rebalanceConsumer: RebalanceConsumer
+  )(commits: Chunk[Commit]): Task[Unit] = {
+    val (offsets, callback, onFailure) = asyncCommitParameters(commits)
+    ZIO.logTrace(s"Async commit of ${offsets.size} offsets for ${commits.size} commits") *>
+      rebalanceConsumer
+        .commitAsync(offsets, callback)
+        .catchAll(onFailure)
+        .unit
```

Comment:
It's already returning `Unit`
```suggestion
      rebalanceConsumer
        .commitAsync(offsets, callback)
        .catchAll(onFailure)
```